win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1055,155 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
     return TRUE;    /* success */
 }
 
+static inline VOID
+FontLink_Chain_Init(PFONTLINK_CHAIN pChain, PTEXTOBJ pTextObj, FT_Face face, PDC dc)
+{
+    RtlZeroMemory(pChain, sizeof(*pChain));
+    pChain->pBaseTextObj = pTextObj;
+    pChain->pDefFace = face;
+    pChain->dc = dc;
+}
+
+static VOID
+FontLink_Chain_Free(PFONTLINK_CHAIN pChain)
+{
+    PLIST_ENTRY Entry;
+    PFONTLINK pLink;
+
+    if (!pChain->szzFontLink[0])
+        return;
+
+    while (!IsListEmpty(&pChain->FontLinkList))
+    {
+        Entry = RemoveHeadList(&pChain->FontLinkList);
+        pLink = CONTAINING_RECORD(Entry, FONTLINK, ListEntry);
+        FontLink_Destroy(pLink);
+    }
+}
+
+static NTSTATUS
+FontLink_Chain_LoadReg(PFONTLINK_CHAIN pChain, LPLOGFONTW pLF)
+{
+    NTSTATUS Status;
+    HKEY hKey;
+    DWORD cbData;
+    static const WCHAR s_szzDefFontLink[MAX_FONTLINK] =
+        L"tahoma.ttf,Tahoma\0"
+        L"msgothic.ttc,MS UI Gothic\0"
+        L"mingliu.ttc,PMingLiU\0"
+        L"simsun.ttc,SimSun\0"
+        L"gulim.ttc,Gulim\0"
+        L"\0";
+    static const WCHAR s_szzDefFixedPitchFontLink[MAX_FONTLINK] =
+        L"cour.ttf,Courier New\0"
+        L"msgothic.ttc,MS Gothic\0"
+        L"mingliu.ttc,MingLiU\0"
+        L"simsun.ttc,NSimSun\0"
+        L"gulim.ttc,GulimChe\0"
+        L"\0";
+
+    pChain->szzFontLink[0] = UNICODE_NULL;
+
+    // Open registry key
+    Status = RegOpenKey(
+        L"\\Registry\\Machine\\Software\\Microsoft\\Windows NT\\CurrentVersion\\FontLink\\SystemLink",
+        &hKey);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    // Load the FontLink entry
+    cbData = sizeof(pChain->szzFontLink);
+    Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, pChain->szzFontLink, &cbData);
+    if (!NT_SUCCESS(Status) && SubstituteFontRecurse(pLF)) // Retry with substituted name
+    {
+        cbData = sizeof(pChain->szzFontLink);
+        Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, pChain->szzFontLink, &cbData);
+    }
+
+    ZwClose(hKey); // Close the registry key
+
+    ASSERT(sizeof(pChain->szzFontLink) == sizeof(s_szzDefFontLink));
+
+    if (!NT_SUCCESS(Status))
+    {
+        if (!(pLF->lfPitchAndFamily & FIXED_PITCH))
+            RtlCopyMemory(pChain->szzFontLink, s_szzDefFontLink, sizeof(pChain->szzFontLink));
+        else
+            RtlCopyMemory(pChain->szzFontLink, s_szzDefFixedPitchFontLink, sizeof(pChain->szzFontLink));
+    }
+
+    // Avoid buffer overrun
```

Comment:
That's not the actual reason.
The real reason why doing the double null-termination is because the data is a multi-string.
The first NUL-termination terminates the last string; and the second NUL-termination terminates the whole array.

Reply:
Improved the comment.